File: tests/test_simulation.py
#! /usr/bin/env python3

import unittest
import random
from unittest.mock import patch, MagicMock

from pysinthe.simulation import (Simulation, Reaction, SpeciesReaction, Bind,
                                 Bridge, SpeciesTracker)


class TestSpeciesTrackerMethods(unittest.TestCase):

    @patch("pysinthe.simulation.Signal")
    def test_increment_species(self, mock_signal):
        tracker = SpeciesTracker()
        tracker.increment_species("reactant1", 1)
        self.assertEqual(tracker.species["reactant1"], 1)
        tracker.increment_species("reactant2", 1)
        self.assertEqual(tracker.species["reactant2"], 1)
        tracker.increment_species("reactant1", 2)
        self.assertEqual(tracker.species["reactant1"], 3)
        self.assertFalse(mock_signal.return_value.fire.called)

        mock_reaction = MagicMock()
        mock_reaction.index = 3
        tracker.add_reaction("reactant2", mock_reaction)
        tracker.increment_species("reactant2", 1)
        mock_signal.return_value.fire.assert_called_with(3)

    def test_add_reaction(self):
        tracker = SpeciesTracker()
        mock_reaction = MagicMock()
        tracker.add_reaction("reaction3", mock_reaction)
        self.assertTrue(
            mock_reaction in tracker.species_reaction_map["reaction3"]
        )
        tracker.add_reaction("reaction4", mock_reaction)
        self.assertTrue(
            mock_reaction in tracker.species_reaction_map["reaction3"]
        )
        self.assertTrue(
            mock_reaction in tracker.species_reaction_map["reaction4"]
        )

    def test_add_polymer(self):
        tracker = SpeciesTracker()
        mock_polymer = MagicMock()
        tracker.add_polymer("promoter1", mock_polymer)
        self.assertTrue(
            mock_polymer in tracker.promoter_polymer_map["promoter1"]
        )
        tracker.add_polymer("promoter2", mock_polymer)
        self.assertTrue(
            mock_polymer in tracker.promoter_polymer_map["promoter2"]
        )
        self.assertTrue(
            mock_polymer in tracker.promoter_polymer_map["promoter1"]
        )


class TestReactionMethods(unittest.TestCase):

    def test_init(self):
        reaction = Reaction()
        self.assertEqual(reaction.index, -1)
        self.assertIsNone(reaction.calculate_propensity())
        self.assertIsNone(reaction.execute())


class TestSpeciesReactionMethods(unittest.TestCase):

    def setUp(self):
        self.tracker = SpeciesTracker()
        self.reaction = SpeciesReaction(self.tracker,
                                        1000,
                                        ["reactant1", "reactant2"],
                                        ["product1", "product2"])

    def test_init(self):
        with self.assertRaises(RuntimeError):
            SpeciesReaction(self.tracker,
                            1000,
                            ["reactant1", "reactant2", "reactant3"],
                            ["product1", "product2"])

        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["reactant1"]
        )
        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["reactant2"]
        )
        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["product1"]
        )
        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["product2"]
        )

    def test_calculate_propensity(self):
        self.tracker.increment_species("reactant1", 2)
        self.tracker.increment_species("reactant2", 3)
        prop = self.reaction.rate_constant * 2 * 3
        self.assertEqual(prop, self.reaction.calculate_propensity())

    def test_execute(self):
        self.setUp()
        self.tracker.increment_species("reactant1", 2)
        self.tracker.increment_species("reactant2", 3)
        self.reaction.execute()
        self.assertEqual(self.tracker.species["reactant1"], 1)
        self.assertEqual(self.tracker.species["reactant2"], 2)
        self.assertEqual(self.tracker.species["product1"], 1)
        self.assertEqual(self.tracker.species["product2"], 1)


class TestBindMethods(unittest.TestCase):

    def setUp(self):
        self.tracker = SpeciesTracker()
        Reaction._CELL_VOLUME = float(8e-15)
        self.reaction = Bind(self.tracker,
                             1000,
                             "promoter1",
                             ["pol1", 10, 20])

    def test_init(self):
        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["promoter1"]
        )
        self.assertTrue(
            self.reaction in self.tracker.species_reaction_map["pol1"]
        )

    def test_calculate_propensity(self):
        self.tracker.increment_species("promoter1", 2)
        self.tracker.increment_species("pol1", 3)
        prop = self.reaction.rate_constant * 2 * 3
        self.assertEqual(prop, self.reaction.calculate_propensity())

    @patch("pysinthe.simulation.Polymerase")
    def test_execute(self, mock_polymerase):
        self.setUp()
        self.tracker.increment_species("promoter1", 2)
        self.tracker.increment_species("pol1", 3)
        mock_polymer = MagicMock()
        mock_polymer.count_uncovered.return_value = 3
        self.tracker.promoter_polymer_map["promoter1"] = [mock_polymer]
        self.reaction.execute()
        self.assertTrue(mock_polymerase.called)
        self.assertTrue(mock_polymer.bind_polymerase.called)
        self.assertEqual(self.tracker.species["promoter1"], 1)
        self.assertEqual(self.tracker.species["pol1"], 2)


class TestBridgeMethods(unittest.TestCase):

    @patch("pysinthe.simulation.Signal")
    def setUp(self, mock_signal):
        self.mock_signal = mock_signal
        self.mock_polymer = MagicMock()
        self.reaction = Bridge(self.mock_polymer)

    def test_calculate_propensity(self):
        self.reaction.calculate_propensity()
        self.assertTrue(self.mock_polymer.calculate_propensity.called)

    def test_execute(self):
        self.reaction.execute()
        self.assertTrue(self.mock_polymer.execute.called)

    def test_update(self):
        self.reaction.update()
        # index will not yet have been set
        self.mock_signal.return_value.fire.assert_called_with(-1)


class TestSimulationMethods(unittest.TestCase):

    def setUp(self):
        self.sim = Simulation()

    def test_register_reaction(self):
        # Add one reaction
        reaction1 = MagicMock()
        reaction1.calculate_propensity.return_value = 0.5
        self.sim.register_reaction(reaction1)
        self.assertEqual(self.sim.alpha_list[0], 0.5)
        self.assertEqual(self.sim.reactions.index(reaction1), 0)
        # Add a second reaction
        reaction2 = MagicMock()
        reaction2.calculate_propensity.return_value = 0.8
        self.sim.register_reaction(reaction2)
        self.assertEqual(self.sim.alpha_list[1], 0.8)
        self.assertEqual(self.sim.reactions.index(reaction2), 1)
        # Make sure that alpha_list is really acting as a cache
        reaction1.calculate_propensity.return_value = 0.1
        self.assertNotEqual(self.sim.alpha_list[0], 0.1)

    def test_register_polymer(self):
        self.setUp()
        polymer = MagicMock()
        mock_promoter = MagicMock()
        mock_promoter.type = "promoter"
        mock_promoter.name = "mypromoter"
        polymer.elements = [mock_promoter]
        self.sim.register_polymer(polymer)
        # Make sure promoter/polymer was added to map
        self.assertEqual(self.sim.tracker.polymers_with_promoter("mypromoter"),
                         [polymer])
        # Make sure bridge reaction was added
        self.assertEqual(self.sim.reactions[0].polymer, polymer)

    def test_initialize_propensity(self):
        # add reactions to simulation as before
        self.setUp()
        self.test_register_reaction()
        # Make sure that cache has not been updated
        self.assertNotEqual(self.sim.alpha_list[0], 0.1)
        # Update cache
        self.sim.initialize_propensity()
        self.assertEqual(self.sim.alpha_list[0], 0.1)
        # reaction2 propensity should not have changed
        self.assertEqual(self.sim.alpha_list[1], 0.8)

    def test_update_propensity(self):
        # add reactions to simulation as before
        self.setUp()
        self.test_register_reaction()
        # Make sure that cache has not been updated
        self.assertNotEqual(self.sim.alpha_list[0], 0.1)
        # Update cache
        self.sim.update_propensity(0)
        self.assertEqual(self.sim.alpha_list[0], 0.1)
        # Update propensity in reaction object
        self.sim.reactions[1].calculate_propensity.return_value = 7
        self.assertNotEqual(self.sim.alpha_list[1], 7)
        self.sim.update_propensity(1)
        self.assertEqual(self.sim.alpha_list[1], 7)

    def test_execute(self):
        self.setUp()
        reaction1 = MagicMock()
        reaction1.calculate_propensity.return_value = 0.5
        self.sim.register_reaction(reaction1)
        reaction2 = MagicMock()
        reaction2.calculate_propensity.return_value = 0.8
        self.sim.register_reaction(reaction2)

        self.sim.initialize_propensity()
        random.seed(17)
        old_iteration = self.sim.iteration
        old_time = self.sim.time
        new_time = old_time + 0.5000911660616504
        self.sim.execute()
        self.assertEqual(self.sim.time, new_time)
        self.assertTrue(reaction2.execute.called)
        self.assertEqual(self.sim.iteration, old_iteration + 1)
        self.assertFalse(reaction1.execute.called)
        self.sim.execute()
        self.assertTrue(reaction1.execute.called)

File: tests/test_simulation.cpp
#include <catch.hpp>

#include "simulation.hpp"
#include "tracker.hpp"

TEST_CASE("SpeciesReaction methods", "[Reaction]") {
  auto reaction = std::make_shared<SpeciesReaction>(
      1000, double(8e-15), std::vector<std::string>{"reactant1", "reactant2"},
      std::vector<std::string>{"product1", "product2"});
  auto &tracker = SpeciesTracker::Instance();
  tracker.Clear();
  tracker.cell_volume(double(8e-15));
  tracker.Register(reaction);
  tracker.Increment("reactant1", 2);
  tracker.Increment("reactant2", 3);

  SECTION("Initialization and registration") {
    REQUIRE_THROWS(SpeciesReaction(
        1000, double(8e-15),
        std::vector<std::string>{"reactant1", "reactant2", "reactant3"},
        std::vector<std::string>{"product1", "product2"}));
    // Make sure that maps have been set up appropriately
    auto reactions = tracker.FindReactions("reactant1");
    REQUIRE(std::find(reactions.begin(), reactions.end(), reaction) !=
            reactions.end());
    reactions = tracker.FindReactions("reactant2");
    REQUIRE(std::find(reactions.begin(), reactions.end(), reaction) !=
            reactions.end());
    reactions = tracker.FindReactions("product1");
    REQUIRE(std::find(reactions.begin(), reactions.end(), reaction) !=
            reactions.end());
    reactions = tracker.FindReactions("product2");
    REQUIRE(std::find(reactions.begin(), reactions.end(), reaction) !=
            reactions.end());
  }

  SECTION("Propensity calculation") {
    REQUIRE((1000 * 2 * 3) / (double(6.0221409e+23) * double(8e-15)) ==
            reaction->CalculatePropensity());
  }

  SECTION("Execution") {
    reaction->Execute();
    REQUIRE(tracker.species("reactant1") == 1);
    REQUIRE(tracker.species("reactant2") == 2);
    REQUIRE(tracker.species("product1") == 1);
    REQUIRE(tracker.species("product1") == 1);
  }
}

TEST_CASE("Bind methods", "[Reaction]") {
  auto &tracker = SpeciesTracker::Instance();
  tracker.Clear();
  tracker.cell_volume(double(8e-15));
  // Set up a polymer
  std::vector<std::string> interactions = {"ecolipol"};
  Promoter::Ptr prom;
  Terminator::Ptr term;
  prom = std::make_shared<Promoter>("p1", 5, 15, interactions);
  std::map<std::string, double> efficiency;
  efficiency["ecolipol"] = 0.6;
  term = std::make_shared<Terminator>("t1", 50, 55, interactions, efficiency);

  std::vector<Element::Ptr> elements;
  elements.push_back(prom);
  elements.push_back(term);
  std::vector<std::string> mask_interactions = {"ecolipol"};
  Mask mask = Mask("test_mask", 50, 100, mask_interactions);

  auto polymer =
      std::make_shared<Polymer>("test_polymer", 1, 100, elements, mask);
  polymer->InitElements();
  // Set up a polymerase
  auto polymerase = Polymerase("ecolipol", 10, 30);
  auto reaction = Bind(1000, double(8e-15), "p1", polymerase);

  SECTION("Calculate propensity") {
    tracker.Increment("p1", 2);
    tracker.Increment("ecolipol", 3);
    REQUIRE((1000 * 2 * 3) / (double(6.0221409e+23) * double(8e-15)) ==
            reaction.CalculatePropensity());
  }

  SECTION("Execution") {
    tracker.Clear();
    tracker.Increment("p1", 1);
    tracker.Increment("ecolipol", 1);
    tracker.Add("p1", polymer);
    reaction.Execute();
    REQUIRE(tracker.species("p1") == 0);
    REQUIRE(tracker.species("ecolipol") == 0);
    REQUIRE(polymer->uncovered("p1") == 0);
    REQUIRE(prom->IsCovered());
  }
}

TEST_CASE("SpeciesTracker methods", "[SpeciesTracker]") {
  auto &tracker = SpeciesTracker::Instance();
  tracker.Clear();
  tracker.cell_volume(double(8e-15));

  SECTION("Increment species") {
    tracker.Increment("reactant1", 1);
    REQUIRE(tracker.species("reactant1") == 1);
    tracker.Increment("reactant2", 1);
    REQUIRE(tracker.species("reactant2") == 1);
    tracker.Increment("reactant1", 2);
    REQUIRE(tracker.species("reactant1") == 3);
  }

  SECTION("Add polymer") {
    Polymer::Ptr polymer;
    tracker.Add("promoter1", polymer);
    auto polymers = tracker.FindPolymers("promoter1");
    REQUIRE(std::find(polymers.begin(), polymers.end(), polymer) !=
            polymers.end());
    tracker.Add("promoter2", polymer);
    polymers = tracker.FindPolymers("promoter2");
    REQUIRE(std::find(polymers.begin(), polymers.end(), polymer) !=
            polymers.end());
  }
}

TEST_CASE("Simulation methods", "[Simulation]") {
  auto sim = std::make_shared<Simulation>();
  auto &tracker = SpeciesTracker::Instance();
  tracker.Clear();
  tracker.cell_volume(double(8e-15));
  tracker.Increment("reactant1", 1);

  SECTION("Register reaction") {
    auto reaction1 = std::make_shared<SpeciesReaction>(
        1.5, double(8e-15), std::vector<std::string>{"reactant1"},
        std::vector<std::string>{"product1"});
    sim->RegisterReaction(reaction1);
    sim->InitPropensity();
    REQUIRE(sim->alpha_sum() == 1.5);
    auto reaction2 = std::make_shared<SpeciesReaction>(
        1.5, double(8e-15), std::vector<std::string>{"reactant1"},
        std::vector<std::string>{"product1"});
    sim->RegisterReaction(reaction2);
    sim->InitPropensity();
    REQUIRE(sim->alpha_sum() == 3);
  }

  SECTION("Register polymer and execute") {
    tracker.Clear();
    tracker.cell_volume(double(8e-15));
    // Set up a polymer
    std::vector<std::string> interactions = {"ecolipol"};
    Promoter::Ptr prom;
    Terminator::Ptr term;
    prom = std::make_shared<Promoter>("p1", 5, 15, interactions);
    std::map<std::string, double> efficiency;
    efficiency["ecolipol"] = 0.6;
    term = std::make_shared<Terminator>("t1", 50, 55, interactions, efficiency);

    std::vector<Element::Ptr> elements;
    elements.push_back(prom);
    elements.push_back(term);
    std::vector<std::string> mask_interactions = {"ecolipol"};
    Mask mask = Mask("test_mask", 50, 100, mask_interactions);

    auto polymer =
        std::make_shared<Polymer>("test_polymer", 1, 100, elements, mask);

    sim->RegisterPolymer(polymer);
    REQUIRE(tracker.FindPolymers("p1")[0] == polymer);
    // bind polymerase and make sure alpha list is updated appropriately
    auto polymerase = Polymerase("ecolipol", 10, 30);
    // Wire up reaction with tracker (TODO: clean this up)
    tracker.Increment("ecolipol", 2);
    tracker.Increment("p1", 1);
    auto reaction =
        std::make_shared<Bind>(1000, double(8e-15), "p1", polymerase);
    tracker.Add("p1", reaction);
    tracker.Add("ecolipol", reaction);
    tracker.Add("p1", polymer);
    sim->RegisterReaction(reaction);
    sim->InitPropensity();
    tracker.propensity_signal_.ConnectMember(sim,
                                             &Simulation::UpdatePropensity);
    sim->Execute();
    REQUIRE(sim->alpha_sum() == 30);
    sim->Execute();
    REQUIRE(sim->alpha_sum() == 30);
    for (int i = 0; i < 20; i++) {
      sim->Execute();
    }
    // Alpha_sum should be slightly greather than 30 now that promoter is
    // re-exposed
    REQUIRE(sim->alpha_sum() > 30);
  }
}